docs/eip/5559.md

```diff
@@ -0,0 +1,447 @@
+
+3. Request the user for an `approval` approving the generated data signer, and finally,
+
+4. Post the calldata to gateway along with signatures `sigData` and `approval`, and the `dataSigner`.
+
+These steps are described in detail below.
+
+#### 1. Generate Data Signer
+The data signer must be generated deterministically from ethereum wallet signatures; see figure below.
+
+![](https://raw.githubusercontent.com/namesys-eth/namesys-ccip-write/main/images/keygen.png)
+
+The deterministic key generation can be implemented concisely in a single unified `keygen()` function as follows.
+
+```js
+/* Pseudo-code for key generation */
+function keygen(
+  username, // CAIP identifier for the blockchain account
+  sigKeygen, // Deterministic signature from wallet
+  password // Optional password
+) {
+  // Calculate input key by hashing signature bytes using SHA256 algorithm
+  let inputKey = sha256(sigKeygen);
+  // Calculate salt for keygen by hashing concatenated username, hashed password and hex-encoded signature using SHA256 algorithm
+  let salt = sha256(`${username}:${sha256(password || "")}:${sigKeygen}`);
+  // Calculate hash key output by feeding input key, salt & username to the HMAC-based key derivation function (HKDF) with dLen = 42
+  let hashKey = hkdf(sha256, inputKey, salt, username, 42);
+  // Calculate and return secp256k1 keypair
+  return secp256k1(hashKey) // Calculate secp256k1 keypair from hash key
+}
+```
+
+This `keygen()` function requires three variables: `username`, `password` and `sigKeygen`. Their definitions are given below.
+
+##### `username`
+CAIP-10 identifier `username` is auto-derived from the connected wallet's checksummed address `wallet` and `chainId`.
+
+```js
+/* CAIP-10 identifier */
+const caip10 = `eip155:${chainId}:${wallet}`
+```
+
+##### `password`
+`password` is an optional private field and it must be prompted from the user by the client; this field allows users to change data signers for a given `username`.
+```js
+/* Secret derived key identifier */ 
+// Clients must prompt the user for this
+const password = 'key1'
+```
+
+##### `sigKeygen`
+The data signer must be derived from the owner or manager keys of a node. Message payload for the required `sigKeygen` must then be formatted as:
+
+```text
+Requesting Signature To Generate Keypair(s)\n\nOrigin: ${username}\nProtocol: ${protocol}\nExtradata: ${extradata}
+```
+
+where the `extradata` is calculated as follows,
+
+```solidity
+// Calculating extradata in keygen signatures
+bytes32 extradata = keccak256(
+    abi.encodePacked(
+        pbkdf2(
+            password, 
+            salt, 
+            iterations
+        ), // Stretch password with PBKDF2
+        wallet
+    )
+)
+```
+
+where `PBKDF2` - with `keccak256(abi.encodePacked(username))` as salt and last 5 hex-nibbles converted to `uint` as the iteration count - is used for brute-force vulnerability protection.
```

Comment:
Perhaps better to fix the iteration count to sufficiently large value?

Reply:
Fixed by https://github.com/ensdomains/docs/pull/224/commits/f50c32183c4aae26c06e7107f4ae8be9ff918601